### PR TITLE
feat: final order issued marker (FPLA-1747)

### PR DESCRIPTION
**Pull request by muyinatech**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-1747

### Change description ###
Ensure that final order can only be issued once for a child.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Comment by muyinatech** [comment]
Child selection screen after some final orders already issued:-
ChildSelection

Child selection screen with one remaining child:-
RemainingChild

Summary screen with one remaining child:-
RemainingChildSummary

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelector.java [hunk 1]
```diff
@@ -16,14 +19,25 @@
     private String childCount = "";
     @Builder.Default
     private List<Integer> selected = new ArrayList<>();
+    @Builder.Default
+    private List<Integer> hidden = new ArrayList<>();
 
-    public void generateChildCount(int max) {
+    public static String generateChildCount(int max) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelector.java:25]: Is there a reason for this to return something now? As well as the method that you added.

Comment (muyinatech) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelector.java:25]: It was previously a void method with a side effect of setting the childCount. I thought it would be cleaner to have a method with a return value.

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelector.java:25]: I think it would be nice to, maybe rename it to setChildCountFromInt or something like that and make sure it is annotated with a json ignore

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerMidEventTest.java [hunk 2]
```diff
@@ -119,6 +120,64 @@ void shouldNotPopulateChildSelectorAndLabelWhenYesIsSelected() {
 
             CaseData caseData = mapper.convertValue(callbackResponse.getData(), CaseData.class);
 
+            assertThat(callbackResponse.getData().get("children_label")).isNull();
+            assertThat(caseData.getChildSelector()).isNull();
+            assertThat(callbackResponse.getData().get("remainingChildIndex")).isNull();
+            assertThat(callbackResponse.getData().get("remainingChild")).isNull();
+            assertThat(callbackResponse.getData().get("otherFinalOrderChildren")).isNull();
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerMidEventTest.java:127]: Do you think we could group all these `callbackResponse.getData().get(..` into `assertThat(callbackResponse.getData()).extracting(....).containsOnlyNulls();`. Would probably work for the childSelector as well

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerMidEventTest.java [hunk 3]
```diff
@@ -119,6 +120,64 @@ void shouldNotPopulateChildSelectorAndLabelWhenYesIsSelected() {
...
+                createChild("Paul", true),
+                createChild("Bill", true));
+
+            AboutToStartOrSubmitCallbackResponse callbackResponse = postMidEvent(
+                buildCaseDetails("No", children), "populate-selector");
+
+            CaseData caseData = mapper.convertValue(callbackResponse.getData(), CaseData.class);
+
+            assertThat(callbackResponse.getData().get("children_label")).isNull();
+            assertThat(caseData.getChildSelector()).isNull();
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerMidEventTest.java:170]: These 2 are repeated below

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderController.java [hunk 4]
```diff
@@ -121,11 +122,23 @@ public AboutToStartOrSubmitCallbackResponse handlePopulateSelectorMidEvent(
...
+                    childrenService.getFinalOrderIssuedChildren(caseData.getAllChildren()));
+            } else {
+                caseDetails.getData()
+                    .put("children_label", childrenService.getChildrenLabel(caseData.getAllChildren()));
+                ChildSelector childSelector = ChildSelector.builder()
+                    .childCount(ChildSelector.generateChildCount(caseData.getAllChildren().size()))
+                    .hidden(ChildSelector.generatedHiddenList(caseData.getAllChildren()))
+                    .build();
+                caseDetails.getData().put("childSelector", childSelector);
+            }
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderController.java:140]: I'm thinking this is getting slightly too complex for a controller. Could it go into the children service?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/json/serializer/ChildSelectorSerializer.java [hunk 5]
```diff
@@ -16,6 +16,7 @@
 @JsonComponent
 public class ChildSelectorSerializer extends JsonSerializer<ChildSelector> {
 
+
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/json/serializer/ChildSelectorSerializer.java:19]: unnecessary change

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ChildrenService.java [hunk 6]
```diff
@@ -38,21 +46,69 @@ public boolean allChildrenHaveFinalOrder(List<Element<Child>> children) {
...
+    public String getFinalOrderIssuedChildren(List<Element<Child>> allChildren) {
+        return allChildren.stream()
+            .filter(child -> YES.getValue().equals(child.getValue().getFinalOrderIssued()))
+            .map(child -> {
+                StringBuilder builder = new StringBuilder();
+                builder.append(child.getValue().getParty().getFullName());
+                if (child.getValue().getFinalOrderIssuedType() != null) {
+                    builder.append(String.format(" - %s issued", child.getValue().getFinalOrderIssuedType()));
+
+                }
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ChildrenService.java:109]: Seeing as this is being used twice, thoughts on have this be a method in the child class, maybe asLabel or something like that?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelectorTest.java [hunk 7]
```diff
@@ -1,23 +1,53 @@
...
+            .party(ChildParty.builder()
+                .firstName("Child")
+                .build());
+
+        if (finalOrderIssued) {
+            builder.finalOrderIssued(YesNo.YES.getValue())
+                .finalOrderIssuedType(GeneratedOrderType.CARE_ORDER.getLabel());
+        }
+
+        return element(builder.build());
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelectorTest.java:50]: Maybe we could use the testChild method from `TestDataHelper`
```suggestion
        Element<Child> child = testChild();
        if (finalOrderIssued) {
            child.getValue().setFinalOrderIssued(YesNo.YES.getValue());
            child.getValue().setFinalOrderIssuedType(GeneratedOrderType.CARE_ORDER.getLabel());
        }
        return child;
```

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelector.java (hunk 1 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/selector/ChildSelector.java:25]: Same for the new method you added, `setHiddenFromChildList` or something like that

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerMidEventTest.java (hunk 3 above)
Comment (muyinatech) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderControllerMidEventTest.java:168]: Result of bad merged, fixed.

**Comment by lewisbirks** [comment]
I have noticed that, since picking this up that the permissions have only been added for admin with close case stuff regarding this event. Will make a separate PR after this to give judges permissions as well seeing as they have permissions for the rest of the event.

File: e2e/tests/hmctsAdministersCaseAfterSubmission_test.js [hunk 8]
```diff
@@ -28,7 +28,7 @@ Scenario('HMCTS admin confirms payment', async (I, caseViewPage) => {
   I.see('Processed payments'); // Test to pass AAT, to make better
 });
 
-Scenario('HMCTS admin enters FamilyMan reference number', async (I, caseViewPage, loginPage, enterFamilyManCaseNumberEventPage) => {
+Scenario('HMCTS admin enters FamilyMan reference number @f', async (I, caseViewPage, loginPage, enterFamilyManCaseNumberEventPage) => {
```
Comment (toireasam) [inline e2e/tests/hmctsAdministersCaseAfterSubmission_test.js:31]: Remove @f please

File: e2e/tests/hmctsAdministersCaseAfterSubmission_test.js [hunk 9]
```diff
@@ -156,7 +156,7 @@ Scenario('HMCTS admin enters hearing details and submits', async (I, caseViewPag
   I.seeInTab(['Hearing 2', 'Judge and Justices\' Legal Adviser', 'Justices\' Legal Adviser\'s full name'], hearingDetails[1].judgeAndLegalAdvisor.legalAdvisorName);
 });
 
-Scenario('HMCTS admin share case with representatives', async (I, caseViewPage, enterRepresentativesEventPage) => {
+Scenario('HMCTS admin share case with representatives @f', async (I, caseViewPage, enterRepresentativesEventPage) => {
```
Comment (toireasam) [inline e2e/tests/hmctsAdministersCaseAfterSubmission_test.js:159]: and again

File: ccd-definition/CaseEventToFields/createOrder.json [hunk 10]
```diff
@@ -25,6 +26,33 @@
...
+    "FieldShowCondition": "orderTypeAndDocument=\"DO_NOT_SHOW\"",
+    "ShowSummaryChangeOption": "N"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "createOrder",
+    "CaseFieldID": "remainingChildIndex",
+    "PageFieldDisplayOrder": 3,
+    "DisplayContext": "READONLY",
```
Comment (toireasam) [inline ccd-definition/CaseEventToFields/createOrder.json:49]: could update page field display order to 4? (although I know we don't display it on the UI so maybe you wanted to keep as 3?

File: ccd-definition/AuthorisationCaseField/court-admin.json [hunk 11]
```diff
@@ -510,6 +510,27 @@
     "UserRole": "caseworker-publiclaw-courtadmin",
     "CRUD": "CRUD"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "remainingChildIndex",
+    "UserRole": "caseworker-publiclaw-courtadmin",
+    "CRUD": "CRUD"
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/court-admin.json:518]: Do we need delete permissions for this one?

File: ccd-definition/AuthorisationCaseField/court-admin.json [hunk 12]
```diff
@@ -510,6 +510,27 @@
...
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "remainingChild",
+    "UserRole": "caseworker-publiclaw-courtadmin",
+    "CRUD": "CRUD"
+  },
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/court-admin.json:526]: Same here do we need delete?

File: ccd-definition/AuthorisationCaseField/court-admin.json [hunk 13]
```diff
@@ -510,6 +510,27 @@
...
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "otherFinalOrderChildren",
+    "UserRole": "caseworker-publiclaw-courtadmin",
+    "CRUD": "CRUD"
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/court-admin.json:532]: And here? (if not then remove for judiciary also) :)

File: e2e/tests/hmctsAdministersCaseAfterSubmission_test.js (hunk 8 above)
Comment (lewisbirks) [reply e2e/tests/hmctsAdministersCaseAfterSubmission_test.js:31]: 🤦

File: ccd-definition/AuthorisationCaseField/court-admin.json (hunk 11 above)
Comment (lewisbirks) [reply ccd-definition/AuthorisationCaseField/court-admin.json:518]: Think it can be removed, if it is only a temporary field you don't usually need `D` permissions

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java [hunk 14]
```diff
@@ -241,24 +241,63 @@ void shouldReturnTrueWhenOrderIsAFinalOrder() {
...
+            orderType = OrderTypeAndDocument.builder().type(SUPERVISION_ORDER).subtype(FINAL).build();
+
+            boolean notAllowFinalOrder = service.shouldNotAllowFinalOrder(orderType, someChildren);
+
+            assertThat(notAllowFinalOrder).isFalse();
+        }
+
+        @Test
+        void shouldReturnFalseWhenNotAllChildrenHaveFinalOrderAndOrderSubTypeIsNotFinal() {
+            orderType = OrderTypeAndDocument.builder().type(SUPERVISION_ORDER).subtype(INTERIM).build();
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java:293]: I think it is fine as is for this one due to how the tests are structured

File: ccd-definition/CaseEventToFields/createOrder.json (hunk 10 above)
Comment (lewisbirks) [reply ccd-definition/CaseEventToFields/createOrder.json:49]: Missed that thanks
